doc/man3/SSL_shutdown.pod

```diff
@@ -21,161 +21,267 @@ SSL_shutdown, SSL_shutdown_ex - shut down a TLS/SSL or QUIC connection
+connection, which remains open.
+
+=head2 Remotely-Initiated Shutdown
+
+If the peer was the first to initiate the shutdown process by sending a
+close_notify alert message, an application will be notified of this as an EOF
+condition when calling
+L<SSL_read(3)> (i.e., L<SSL_read(3)> will fail and L<SSL_get_error(3)> will
+return B<SSL_ERROR_ZERO_RETURN>), after all application data sent by the peer
+prior to initiating the shutdown has been read. An application should handle
+this condition by calling SSL_shutdown() to respond with a close_notify alert in
+turn, completing the shutdown process, though it may choose to write additional
+application data using L<SSL_write(3)> before doing so. If an application does
+not call SSL_shutdown() in this case, a close_notify alert will not be sent and
+the behaviour will not be fully standards compliant.
+
+=head2 Shutdown Lifecycle
+
+Regardless of whether a shutdown was initiated locally or by the peer, if the
+underlying BIO is blocking, a call to SSL_shutdown() will return firstly once a
+close_notify alert message is written to the peer (returning 0), and upon a
+second and subsequent call, once a corresponding message is received from the
+peer (returning 1 and completing the shutdown process). Calls to SSL_shutdown()
+with a blocking underlying BIO will also return if an error occurs.
+
+If the underlying BIO is nonblocking and the shutdown process is not yet
+complete (for example, because a close_notify alert message has not yet been
+received from the peer, or because a close_notify alert message needs to be sent
+but would currently block), SSL_shutdown() returns 0 to indicate that the
+shutdown process is still ongoing; in this case, a call to L<SSL_get_error(3)>
+will yield B<SSL_ERROR_WANT_READ> or B<SSL_ERROR_WANT_WRITE>.
+
+An application can then detect completion of the shutdown process by calling
+SSL_shutdown() again repeatedly until it returns 1, indicating that the shutdown
+process is complete (with a close_notify alert having both been sent and
+received).
+
+However, the preferred method of waiting for the shutdown to complete is to use
+L<SSL_read(3)> until L<SSL_get_error(3)> indicates EOF by returning
+B<SSL_ERROR_ZERO_RETURN>. This ensures any data received immediately before the
+peer's close_notify alert is still provided to the application. It also ensures
+any final handshake-layer messages received are processed (for example, messages
+issuing new session tickets).
+
+If this approach is not used, the second call to SSL_shutdown() (to complete the
+shutdown by confirming receipt of the peer's close_notify message) will fail if
+it is called when the application has not read all pending application data
+sent by the peer using L<SSL_read(3)>.
+
+SSL_shutdown() should not be called if a previous fatal error has occurred on a
+connection; i.e., if L<SSL_get_error(3)> has returned B<SSL_ERROR_SYSCALL> or
+B<SSL_ERROR_SSL>.
```

Comment:
This paragraph seems like it should also be relevant to QUIC

Reply:
Is it? While it may not succeed it is not going to do any harm with the QUIC Implementation.

Reply:
No, but I consider this part of the API contract for SSL_shutdown() regardless of the underlying protocol. So I don't see this as TLS/DTLS specific advice.

Reply:
OK